### PR TITLE
[TIC-94 & 93] Admin Purchase Page backend implementation & front end updates

### PR DESCRIPTION
**Pull request by 0Tendo**
### Description
Implemented backend for Admin ticket purchasing and ticket processing.
Validation for ticket qty vs available sets & complete selection fields
Updated buttons and popups to match the rest of site

image
image

### Risks
May have merge conflicts or require updates depending on changes for related ticket processing tickets:

### Validation
Manual testing of components and entire admin purchasing story

### Issue
N/A

### Operating System
Windows

### CODE REVIEW

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCart.tsx [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import React, {ReactElement, useState, useEffect} from 'react';
+import {useLocation, useNavigate} from 'react-router';
+import {EventRow} from './AdminPurchase';
+
+/**
+ * Math to dollar - `$${(Math.round(x * 100) / 100).toFixed(2)}`
+ *
+ * @param x
+ */
+const toDollar = (x: number) => `$${(Math.floor(x * 100) / 100).toFixed(2)}`;
```
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCart.tsx:10]: Why is this necessary. Does just doing `toFixed(2)` not work?

File: client/src/components/Ticketing/checkout/CheckoutSuccess.tsx [hunk 2]
```diff
@@ -10,28 +10,46 @@
...
             </svg>
-      Thank you for your purchase!
+            Thank you for your purchase!
           </div>
-          <button onClick={() => navigate('/')} className="bg-blue-500 px-5 py-3 text-white text-lg font-normal rounded-xl hover:bg-blue-600"> Back to Home</button>
+          <button
+            onClick={() => navigate('/')}
+            className='bg-blue-500 px-5 py-3 text-white text-lg font-normal rounded-xl hover:bg-blue-600'
+          >
+            {' '}
```
Comment (henrykaus) [inline client/src/components/Ticketing/checkout/CheckoutSuccess.tsx:49]: Instead of adding a space here, you could add a margin-right to the `div` element just before the `button`

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCart.tsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  const itemsInCart = cartItems.map((item, index) => (
+    <div
+      key={index}
+      className='flex flex-row gap-8 bg-gradient-to-b from-zinc-700 px-5 py-3 rounded-xl mb-5'
+    >
+      <div className='flex-auto text-left'>
+        {item.qty} x {item.name}
+      </div>
+      <div className='flex-auto text-right'>{toDollar(item.price)}</div>
+    </div>
```
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCart.tsx:39]: To give the order items more semantic meaning, I would recommend making the outer div a `p` element with two `span`s within rather than 2 divs. This signals to any screen-reader, as well as just browsers in general, that this is textual information.

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCart.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+          Add more items
+        </button>
+      </div>
+      <div className='flex flex-col items-center gap-2 bg-zinc-800 rounded-xl px-5 py-3'>
+        <div className='flex flex-row items-center gap-2 justify-between w-full'>
+          <div className='text-zinc-100 text-sm '>Total</div>
+          <div className='text-white text-lg font-bold'>{toDollar(total)}</div>
+        </div>
+      </div>
+    </div>
```
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCart.tsx:68]: I would recommend using `div`s a little less frequently, as there are many html elements that are able to provide more information and do not affect the visual look. Because of the way wondertix is set up, there isn't much visual difference between `div`s, `button`s, `p`s, `a`s, etc. My recommendation for this piece of code would be to use a `section` as the parent element, with an `h2` as the Your Order. (an h2 instead of an h1 as there should only be 1 h1 per page but there can be several h2s). And then using more `p` elements for text

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx [hunk 5]
```diff
@@ -22,62 +24,64 @@ const stripePromise = loadStripe(pk);
...
+  console.log('eventDataFromPurchase:', eventDataFromPurchase);
   const navigate = useNavigate();
-  const cartItems = useAppSelector(selectCartContents);
+  const cartItems = location.state?.cartItems || [];
   const discount = useAppSelector(selectDiscount);
-  const donation = useAppSelector(selectDonation);
   const handleBackButton = () => {
     navigate('/ticketing/purchaseticket');
   };
+  const dispatch = useAppDispatch();
```
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx:37]: Could you organize this section of code into three parts: hooks, variables with state, and then put the handleBackButton function last. This should make the code a bit easier to decipher. Also we can also remove the `console.log`

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx [hunk 6]
```diff
@@ -22,62 +24,64 @@ const stripePromise = loadStripe(pk);
...
-        method: 'POST',
-        headers: {
-          'Content-Type': 'application/json',
+    try {
+      if (formData.seatingAcc === 'Other') {
+        formData.seatingAcc = formData.comments;
+      }
+      const donation = formData.donation;
+      const stripe = await stripePromise;
+      if (!stripe) return;
```
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx:46]: Could there be a newline below for separation of parts

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx [hunk 7]
```diff
@@ -22,62 +24,64 @@ const stripePromise = loadStripe(pk);
...
-      console.error(result.error.message);
+      );
+      if (!response.ok) {
+        throw response;
+      }
+      const session = await response.json();
+      if (session.id === 'comp') {
+        dispatch(removeAllTicketsFromCart());
+        navigate(`/success`);
+        // return <CheckoutSuccess />;
```
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx:65]: Should this be removed or uncommented?

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx [hunk 8]
```diff
@@ -22,62 +24,64 @@ const stripePromise = loadStripe(pk);
...
-                fillRule='evenodd'
-                d='M9.707 14.707a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414l4-4a1 1 0 011.414 1.414L7.414 9H15a1 1 0 110 2H7.414l2.293 2.293a1 1 0 010 1.414z'
-                clipRule='evenodd'
-              />
-            </svg>
-            back to Purchase
-          </button>
-        </div>
+        <div className='flex flex-row items-center h-auto mt-[7.25rem] md:w-full mb-5'></div>
         <div className='flex flex-row items-center mt-2 text-zinc-800'></div>
```
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx:85]: These 2 divs can be removed

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx [hunk 9]
```diff
@@ -22,62 +24,64 @@ const stripePromise = loadStripe(pk);
...
         <div className='flex flex-col items-center md:flex-row rounded-[1rem] md:items-stretch md:bg-white sm:flex-col w-full h-full'>
```
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx:86]: When the 2 divs are removed, this div will need the `mt-[7.5rem]`

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCompleteOrderForm.tsx [hunk 10]
```diff
@@ -29,35 +32,39 @@ export interface CheckoutFormInfo {
...
  * @returns {ReactElement}
  */
-export default function CompleteOrderForm({
+export default function AdminCompleteOrderForm({
   onSubmit,
-  onBack,
+  backButtonRoute,
+  eventDataFromPurchase,
   disabled,
   donationForm,
```
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCompleteOrderForm.tsx:65]: Since these props don't do anything, they can be removed

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCompleteOrderForm.tsx [hunk 11]
```diff
@@ -29,35 +32,39 @@ export interface CheckoutFormInfo {
...
  * @param {boolean} disabled
  * @param {boolean} donationForm
  */
 type CompleteOrderFormProps = {
   onSubmit: (formData: CheckoutFormInfo) => any;
-  onBack: () => any;
+  backButtonRoute: string;
+  eventDataFromPurchase: EventRow[];
   disabled: boolean;
   donationForm?: boolean;
```
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCompleteOrderForm.tsx:50]: They'd be removed here too

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx [hunk 12]
```diff
@@ -86,9 +90,10 @@ export default function CheckoutPage(): ReactElement {
                 Complete Order
               </div>
               <AdminCompleteOrderForm
-                disabled={cartItems.length === 0}
+                disabled={false} // {cartItems.length === 0}
```
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx:93]: I don't think the `disabled` prop does anything so it can be removed

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCompleteOrderForm.tsx [hunk 13]
```diff
@@ -69,7 +76,7 @@ export default function CompleteOrderForm({
   const [seatingAcc, setseatingAcc] = useState('');
   const [comments, setComments] = useState('');
   const [optIn, setOptIn] = useState(false);
-  const [donation, setDonations] = useState(''); // TODO: may need to add donation to formData if going this route
+  const [donation, setDonations] = useState<number>(0); // TODO: may need to add donation to formData if going this route
```
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCompleteOrderForm.tsx:79]: What does this TODO mean?

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminPurchase.tsx [hunk 14]
```diff
@@ -165,13 +212,13 @@ const AdminPurchase = () => {
       headerName: '',
       width: 150,
       renderCell: (params) => (
-        <Button
-          variant='contained'
-          color='secondary'
+        <button
+          className='bg-red-500 px-2 py-1 text-white rounded-xl hover:bg-red-600 disabled:opacity-40 m-2'
+          type='submit'
```
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminPurchase.tsx:217]: Is `type='submit'` necessary?

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminPurchase.tsx [hunk 15]
```diff
@@ -184,23 +231,28 @@ const AdminPurchase = () => {
         );
         const jsonRes = await response.json();
         const jsonData = jsonRes.data as any[];
+        console.log('API Response for Events:', jsonData);
```
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminPurchase.tsx:234]: Can this be removed

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminPurchase.tsx [hunk 16]
```diff
@@ -286,174 +343,238 @@ const AdminPurchase = () => {
...
     // Extract the numerical value of the price
     const price = parseFloat(selectedType?.price.replace(/[^\d.-]/g, '')) || 0;
 
-    // Update the eventData
+    let seatsForType;
+
+    if (row.eventinstanceid) {
+      try {
+        const response = await fetch(
+          `${process.env.REACT_APP_API_1_URL}/tickets/restrictions/`,
```
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminPurchase.tsx:368]: I don't think you need the trailing `/`

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminPurchase.tsx [hunk 17]
```diff
@@ -286,174 +343,238 @@ const AdminPurchase = () => {
...
+        !row.eventtime ||
+        !row.ticketTypes ||
+        typeof row.price === 'undefined'
+      ) {
+        setDialog(true);
+        setErrMsg('Missing selection.');
+        return;
+      }
+    }
+    console.log('eventData', eventData);
```
Comment (henrykaus) [inline client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminPurchase.tsx:475]: We don't need this

**Review by henrykaus — CHANGES_REQUESTED** [review]
I noticed one issue when testing this branch. When you remove all items from the `Purchase Tickets` screen, then you are able to go to the Checkout Form screen with no items. I left a bunch of comments, but honestly most of them are nitpicks, this looks really good!

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminPurchase.tsx (hunk 14 above)
Comment (0Tendo) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminPurchase.tsx:217]: nope! removed.

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCart.tsx (hunk 1 above)
Comment (txiong002) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCart.tsx:10]: Changed to x.toFixed(2).  Tested no issues.
image
image

File: client/src/components/Ticketing/checkout/CheckoutSuccess.tsx (hunk 2 above)
Comment (txiong002) [reply client/src/components/Ticketing/checkout/CheckoutSuccess.tsx:49]: Removed the space.

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCart.tsx (hunk 3 above)
Comment (txiong002) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCart.tsx:39]: replaced div's with p and span tags. tested no issues.
image

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCart.tsx (hunk 4 above)
Comment (txiong002) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCart.tsx:68]: replaced some div's with better information tags. Tested no issues.
image

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx (hunk 5 above)
Comment (txiong002) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx:37]: organized variables and function.
image

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx (hunk 7 above)
Comment (txiong002) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx:65]: I'll remove it for now.  I was trying to fix the back button route for the success page

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx (hunk 8 above)
Comment (txiong002) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx:85]: removed divs

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx (hunk 9 above)
Comment (txiong002) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx:86]: add mt-[6rem] to div
image

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCompleteOrderForm.tsx (hunk 10 above)
Comment (txiong002) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCompleteOrderForm.tsx:65]: removed

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCompleteOrderForm.tsx (hunk 11 above)
Comment (txiong002) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCompleteOrderForm.tsx:50]: removed

File: client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx (hunk 12 above)
Comment (txiong002) [reply client/src/components/Ticketing/ticketingmanager/AdminPurchase/AdminCheckout.tsx:93]: removed

**Review by henrykaus — APPROVED** [review]

**Review by henrykaus — COMMENTED** [review]
I'll approve, but because we skipped the checkout process, there is no way to "refund" an order of comped tickets, meaning no way to get those tickets back once they are gone. This might not be a big deal though. Just a scenario.

**Comment by benfaustin95** [comment]
> I'll approve, but because we skipped the checkout process, there is no way to "refund" an order of comped tickets, meaning no way to get those tickets back once they are gone. This might not be a big deal though. Just a scenario.

That's my oversight I can fix the refund process.
